crates/ilp-cli/src/interpreter.rs

```
@@ -142,8 +144,9 @@ impl NodeClient<'_> {
             _ => Err(Error::ProtocolErr("Unexpected protocol")),
         }?;
 
-        // The scheme has already been sanitized so this should always succeed
-        url.set_scheme(scheme).expect("Could not alter URL scheme");
+        if let Err(_) = url.set_scheme(scheme) {
+            return Err(Error::SchemeErr());
+        };
```

Curious about the choice to do this approach rather than `url.set_scheme(scheme)?;`, is it to avoid needing to have a new `From` impl? If that's your concern then this may be a good excuse to take a look at [`thiserror`](https://github.com/dtolnay/thiserror); check out the `#[from]` attribute it provides.

---

Good question.

The goal was to return an error with a specific message. The challenge is that `set_scheme` returns a pretty generic `Result<(), ()>`. I was concerned that using `impl From<()>` with `?` would frustrate future developers if every `Error<()>` is automatically converted into to a `SchemeErr` with a very specific failure message. Alternatively, we could return a generic error message but this didn't seem user-friendly. Or something I don't yet know about?

I don't mind switching to `thiserror`, but do you think we should switch the project as a whole or an individual create?

Also, I need to run clippy ;)

---

I would accept either having the `From` here (I'm not too worried about future developers running into problems at this point, and maybe in the meantime we can convince the url crate to use a proper error type here).

> I don't mind switching to thiserror, but do you think we should switch the project as a whole or an individual create?

Just this crate, and only if you like testing out new libraries as much as I do. :)

---

What behavior do we want if `rust-url` updates their API?
If we implement `From` on `()`, our code will break when their API changes.
If we use `is_err()` and manually create an error, then an API change will slip by unnoticed.
Both have their benefits.

It looks like this might not happen for some time, as you saw in this [issue](https://github.com/servo/rust-url/issues/299) and [PR](https://github.com/servo/rust-url/pull/515).

Re `failure` vs `thiserror`, I'm tempted to merge using `failure` and resurrect the question after exprimenting a bit. :)

---

> I'm tempted to merge using failure

Sounds good to me!